File: DotNetPoly/Player/HumanPlayer.vb
Imports DotNetPoly.Fields

Public Class HumanPlayer
    Inherits BasePlayer


    Sub New(pName As String, pColor As System.Drawing.Color)
        MyBase.New(pName, pColor)
    End Sub

    Overrides Function DelegateDiceRoll() As DiceRollResult
        RaiseEvent DelegateControl(Me, New eActionType() {eActionType.DiceOnly}, Nothing, Nothing, 0)
        WaitChoosenAction.WaitOne()

        WaitChoosenAction.Reset()

        Return onRollDice()
    End Function

    Friend Overrides Sub onDelegateControl(pPossibleActions() As eActionType, pfields() As BaseField, pPlayers() As Entity, pCash As Integer)
        Dim choosenPossibility As eActionType
        Dim allPossibilities As New Collections.Generic.List(Of eActionType)(3)

        allPossibilities.AddRange(pPossibleActions)
        allPossibilities.Add(eActionType.GiveUp)

        If allPossibilities.Contains(eActionType.PayRent) Then
            If GameBoard.FreeParkingOwner.Equals(Me) Then
                allPossibilities.Add(eActionType.CardUseFreeParking)
            End If
        End If

        If allPossibilities.Contains(eActionType.JumpToJail) Then
            If GameBoard.FreeJailOwner.Equals(Me) Then
                allPossibilities.Add(eActionType.CardUseFreeJail)
            End If
        End If

        RaiseEvent DelegateControl(Me, allPossibilities.ToArray, pfields, pPlayers, pCash)

        WaitChoosenAction.WaitOne()

        WaitChoosenAction.Reset()

        Try
            choosenPossibility = allPossibilities.ToArray(choosenActionIndex - 1)
        Catch ex As Exception
            Console.WriteLine("choosenActionIndex {0} is not in list {1}", choosenActionIndex, String.Join(",", allPossibilities))
            choosenPossibility = allPossibilities.ToArray(0)
        End Try

        invokeAction(choosenPossibility, pfields, pPlayers, pCash)
    End Sub

    Overrides Sub SubmitChoosenAction(pIndex As Integer, pObject As Object)
        choosenActionIndex = pIndex
        WaitChoosenAction.Set()
    End Sub

    Event DelegateControl(pPlayer As BasePlayer, pPossibleActions() As eActionType, pFields As BaseField(), pPlayer() As Entity, pCash As Integer)


End Class

File: DotNetPoly/Player/HumanPlayer.cs
using System.Linq;

namespace DotNetPoly
{
    public class HumanPlayer : Player
    {
        public HumanPlayer(string pName, System.Drawing.Color pColor) : base(pName, pColor) { }

        public override DiceRollResult DelegateDiceRoll()
        {
            var x = new DelegateEventArgs(new eActionType[] { eActionType.DiceOnly }, null, null, 0);
            DelegateControl?.Invoke(this, ref x);
            WaitChoosenAction.WaitOne();

            WaitChoosenAction.Reset();

            return onRollDice();
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="e"></param>
        internal override void onDelegateControl(DelegateEventArgs e)
        {
            System.Collections.Generic.List<eActionType> allPossibilities = new System.Collections.Generic.List<eActionType>(3);

            // Keep e for recall
            var modifiedEventArgs = e.Copy();

            allPossibilities.AddRange(modifiedEventArgs.Actions);


            if (allPossibilities.Contains(eActionType.PayRent))
            {
                if (CheatAlwaysRentFree || GameBoard.FreeParkingOwner.Equals(this))
                    allPossibilities.Add(eActionType.Card_UseFreeParking);
            }

            if (allPossibilities.Contains(eActionType.JumpToJail))
            {
                if (CheatAlwaysJailFree || GameBoard.FreeJailOwner.Equals(this))
                    allPossibilities.Add(eActionType.Card_UseFreeJail);
            }

            if (GameBoard.Settings.FixedDepositePercentage > 0)
            {
                if (e.Cash > 0 && !allPossibilities.Contains(eActionType.LoseTurn))
                    allPossibilities.Add(eActionType.IncreaseFixedDeposite);

                if (Festgeld > 0 && !allPossibilities.Contains(eActionType.LoseTurn))
                    allPossibilities.Add(eActionType.TerminateFixedDeposite);
            }

            if (GameBoard.Settings.UseCredits)
            {
                if (GameBoard.Statistics.SoldHouseFields > 0 && this._ownfields.Count > 0 && !allPossibilities.Contains(eActionType.LoseTurn))
                    if ((from h in OwnFields where !h.Value.HasCredit select h).Any())
                        allPossibilities.Add(eActionType.CreditForHouse);

                if (GameBoard.Statistics.CreditHouses > 0)
                {
                    var hasCredit = (from h in OwnFields where h.Value.HasCredit select h).Any();

                    if (hasCredit)
                        allPossibilities.Add(eActionType.CreditReleaseHouse);
                }
            }


            allPossibilities.Add(eActionType.GiveUp);

            modifiedEventArgs.Actions = allPossibilities.ToArray();
            DelegateControl?.Invoke(this, ref modifiedEventArgs);

            WaitChoosenAction.WaitOne();

            WaitChoosenAction.Reset();

            // TODO: Next: Check if choosen vars are fileld

            if (invokeAction(choosenEventArgs))
                onDelegateControl(e);
        }

        // Event DelegateControl(pPlayer As BasePlayer, pPossibleActions() As eActionType, pFields As BaseField(), pPlayer() As Entity, pCash As UInteger)
        public event DelegateControlEventHandler DelegateControl;

        public delegate void DelegateControlEventHandler(Player pPlayer, ref DelegateEventArgs e);
    }
}
